midenc-compile/src/stages/cross_ctx/lift_exports.rs

```diff
@@ -0,0 +1,170 @@
+//! lifting the exports for cross-context calls.
+
+use std::collections::BTreeMap;
+
+use miden_assembly::Spanned;
+use midenc_hir::{
+    diagnostics::Severity,
+    pass::AnalysisManager,
+    types::Abi::{self, Canonical},
+    ComponentBuilder, ComponentExport, FunctionType, InstBuilder, InterfaceFunctionIdent,
+};
+use midenc_session::{DiagnosticsHandler, Session};
+
+use super::flat::{
+    assert_core_wasm_signature_equivalence, flatten_function_type, needs_transformation,
+    FlatteningDirection,
+};
+use crate::{stage::Stage, CompilerResult, LinkerInput};
+
+/// Generates lifting for exports for the cross-context calls according to the Miden ABI.
+///
+/// For each component export ensure a module for each interface that has a lifting
+/// function, i.e. a function that takes parameters according to the cross-context Miden
+/// ABI, converts them to the Wasm CABI, calls the core Wasm module exported function,
+/// converts the results to the cross-context Miden ABI
+///
+/// After this stage all exported functons are expected to be called using the Miden ABI for
+/// cross-context calls, i.e. using the stack and the advice provider for arguments and results.
+pub struct LiftExportsCrossCtxStage;
+
+impl Stage for LiftExportsCrossCtxStage {
+    type Input = LinkerInput;
+    type Output = LinkerInput;
+
+    fn run(
+        &mut self,
+        input: Self::Input,
+        _analyses: &mut AnalysisManager,
+        session: &Session,
+    ) -> CompilerResult<Self::Output> {
+        let component = if let LinkerInput::Hir(component) = input {
+            component
+        } else {
+            return Ok(input);
+        };
+
+        // dbg!(&component.exports());
+        // dbg!(&component.modules().keys());
+
+        let mut component_builder = ComponentBuilder::load(*component, &session.diagnostics);
+
+        let mut lifted_exports: BTreeMap<InterfaceFunctionIdent, ComponentExport> = BTreeMap::new();
+        let exports = component_builder.exports().clone();
```

Comment:
*This note applies only to integrating HIR2 with the work in this PR:*

I would expect that the only exports from a component are either component functions (i.e. `CanonLift` ABI, from my other comment), or interfaces comprised of such functions. It looks like the way we're building components in the current IR is flattening the relationship between the component and the underlying module that provides the implementation of the component exports. Mostly just want to note that we'll need to make sure that when translating to HIR2 that we preserve more of the original component structure and visibilities.